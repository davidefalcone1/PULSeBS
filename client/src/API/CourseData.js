//usata per la retrieve della lista dei corsi di un docente
class CourseData{
    constructor(courseId, courseName, teacherId, 
        normalBookingsAvgWeek, cancelledBookingsAvgWeek, waitingBookingsAvgWeek, 
        normalBookingsAvgMonth, cancelledBookingsAvgMonth, waitingBookingsAvgMonth,
        attendanceCountAvgMonth, attendanceCountAvgWeek ) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
        this.normalBookingsAvgWeek = normalBookingsAvgWeek;
        this.cancelledBookingsAvgWeek = cancelledBookingsAvgWeek;
        this.waitingBookingsAvgWeek = waitingBookingsAvgWeek;
        this.normalBookingsAvgMonth = normalBookingsAvgMonth;
        this.cancelledBookingsAvgMonth = cancelledBookingsAvgMonth;
        this.waitingBookingsAvgMonth = waitingBookingsAvgMonth;
        this.attendanceCountAvgWeek = attendanceCountAvgWeek;
        this.attendanceCountAvgMonth = attendanceCountAvgMonth;
    }

    static fromJson(json){
        const temp =  Object.assign(new CourseData(), json);
        return temp;
    }
}

export default CourseData;