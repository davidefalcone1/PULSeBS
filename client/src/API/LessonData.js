import moment from 'moment';

class LessonData {
    constructor(scheduleId, courseId, startDate, endDate, occupiedSeats, availableSeats,
        isLessonCancelled, isLessonRemote, classroom
        /*, normalBookings, cancelledBookings, waitingBookings */) {
        if (scheduleId)
            this.scheduleId = scheduleId;
        this.courseId = courseId;
        this.startDate = moment(new Date(startDate));
        this.endDate = moment(new Date(endDate));
        this.occupiedSeats = occupiedSeats;
        this.availableSeats = availableSeats;
        this.isLessonCancelled = isLessonCancelled;
        this.isLessonRemote = isLessonRemote;
        this.classroom = classroom

        //TODO use the correct variables
        this.normalBookings = 0;
        this.cancelledBookings = 0;
        this.waitingBookings = 0;
    }

    static fromJson(json) {
        const temp = Object.assign(new LessonData(), json);
        temp.startDate = moment(new Date(temp.startDate));
        temp.endDate = moment(new Date(temp.endDate));
        return temp;
    }
}

export default LessonData;