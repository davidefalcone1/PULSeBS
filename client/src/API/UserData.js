class UserData{
    constructor(id, personId, fullName, email, hasDoneTutorial) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
        this.hasDoneTutorial = hasDoneTutorial;
    }

    static fromJson(json){
        const temp =  Object.assign(new UserData(), json);
        return temp;
    }
}

export default UserData;