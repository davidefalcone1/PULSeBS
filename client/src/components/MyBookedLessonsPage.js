import React from 'react';
import LessonListItem from './LessonListItem';
import ListGroup from 'react-bootstrap/ListGroup';
import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button'
import Accordion from 'react-bootstrap/Accordion';
import Card from 'react-bootstrap/Card';
import Form from 'react-bootstrap/Form';
import Col from 'react-bootstrap/Col';
import moment from 'moment';
import { Redirect } from 'react-router-dom';
import { AuthContext } from '../_services/AuthContext';

const lessonsList = (props) => {
  return(
    <LessonListPageRender coursesList={props.courses}
        lessonsList={props.lessonsList}
        waitingBookings={props.waitingBookings}
        selectLessonFunction={props.selectLessonFunction}
        updateMyBookedLessonsList={props.updateMyBookedLessonsList}/>
  );
}

class LessonListPageRender extends React.Component {

  constructor(props) {
      super(props);
      this.props = props;
      this.state = {
        bookingMessage: undefined, bookingCompleted:false,
        startingDay: '', endingDay: '', courseName: 'Select course name'
      }
  }

  updateModalMessage = (msg) => {
    this.setState({bookingMessage: msg, bookingCompleted: true});
  }

  updateField = (name, value) => {
    this.setState({[name]: value}, () => {
      if(name === 'startingDay' && moment(this.state.startingDay).isAfter(moment(this.state.endingDay))){
        this.setState({endingDay: ''}, () => {})
      }
    });
  }
  resetFields = () => {
    this.setState({startingDay: '', endingDay: '', courseName: 'Select course name'}, () => {});
  }
  isLessonInDateBoundaries = (lessonStartingDay, lessonEndingDay) => {
    if(this.state.startingDay === '' && this.state.endingDay === ''){
      return true;
    }
    else if(this.state.startingDay !== '' 
      && moment(this.state.startingDay).isSameOrBefore(moment(lessonStartingDay)) 
      && this.state.endingDay === ''){
        return true;
      }
    else if(this.state.startingDay !== '' 
      && moment(this.state.startingDay).isSameOrBefore(moment(lessonStartingDay)) 
      && this.state.endingDay !== ''
      && moment(this.state.endingDay).isSameOrAfter(moment(lessonEndingDay)) ){
        return true;
    }
    else if(this.state.startingDay === '' 
      && this.state.endingDay !== ''
      && moment(this.state.endingDay).isSameOrAfter(moment(lessonEndingDay)) ){
        return true;
    }
    else{
      return false;
    }
  }
  isCourseSelected = (currentCourseName) => {
    if(this.state.courseName === 'Select course name'){
      return true;
    }
    else if (this.state.courseName === currentCourseName){
      return true;
    }
    else{
      return false;
    }
  }

  render(){
    return(
      <AuthContext.Consumer>
        {(context) => (
          <>        
            {context.user && (this.props.lessonsList || this.props.waitingBookings )&& 
            <div style={{padding: "15px"}}>
              <h5>Filter your lessons!</h5>
              <Form method="POST" action="" id="lessonFilterForm" onSubmit={(ev) => {
                        ev.preventDefault();
                }} ref={(form) => this.form = form}>
                <Form.Row>
                  <Form.Group as={Col}>
                    <Form.Label className="control-label">Course Name</Form.Label>
                    <Form.Control as="select" custom name="courseName" value = {this.state.courseName} 
                        onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                      <option>Select course name</option>
                      {this.props.coursesList.map((course) =>
                        <option>{course.courseName}</option>
                      )}
                    </Form.Control>
                  </Form.Group>
                  <Form.Group as={Col}>
                    <Form.Label className="control-label">Starting Day</Form.Label>
                    <Form.Control type="date" name="startingDay"
                      value = {this.state.startingDay} min={moment().format("YYYY-MM-DD")}
                      onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                  </Form.Group>
                  <Form.Group as={Col}>
                    <Form.Label className="control-label">Ending day</Form.Label>
                    <Form.Control type="date" name="endingDay"
                      value = {this.state.endingDay} min={moment(this.state.startingDay).format("YYYY-MM-DD")}
                      onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                  </Form.Group>
                </Form.Row>

                <Form.Group>
                  <div>
                    <button type="button" className="btn btn-secondary" onClick={() => this.resetFields()}>Reset Fields</button>
                  </div>
                </Form.Group>                  
              </Form>
                    
              <br/>
              <Accordion>
                {this.props.lessonsList &&
                  <Card>
                    <Card.Header>
                      <Accordion.Toggle as={Button} variant="link" eventKey = {"bookedLessons"}>
                        <div className="d-flex w-100 pt-3 justify-content-between no-gutters" id = {"bookedLessons"}>
                            <h3>Booked Lessons</h3>
                        </div>
                      </Accordion.Toggle>
                    </Card.Header>
                    <Accordion.Collapse eventKey={"bookedLessons"}>
                      <Card.Body>
                        <Accordion>
                          {this.props.coursesList.map((course) => //per ogni mio corso
                            (this.isCourseSelected(course.courseName)) &&
                            <Card>
                              <Card.Header>
                                <Accordion.Toggle as={Button} variant="link" eventKey = {"course-" + course.courseName}>
                                  <CourseHeader course = {course.courseName}/>
                                </Accordion.Toggle>
                              </Card.Header>
                              <Accordion.Collapse eventKey={"course-" + course.courseName}>
                                <Card.Body>
                                  <ListGroup as="ul" variant="flush">
                                    <ListHeader />
                                    {this.props.lessonsList.map((lesson) => 
                                      (course.courseId === lesson.courseId) &&
                                      (this.isLessonInDateBoundaries(lesson.startingTime, lesson.endingTime)) &&
                                      <LessonListItem key = {lesson.scheduleId} lesson = {lesson}
                                        updateSelectionMessage = {this.updateSelectionMessage}
                                        updateLessonSelectedState = {this.updateLessonSelectedState}
                                        selectLessonFunction = {this.props.selectLessonFunction}
                                        updateModalMessage={this.updateModalMessage}/>
                                    )}
                                  </ListGroup>
                                </Card.Body>
                              </Accordion.Collapse>
                            </Card>
                          )}
                        </Accordion>
                      </Card.Body>
                    </Accordion.Collapse>
                  </Card>
                }
                {this.props.waitingBookings &&
                  <Card>
                    <Card.Header>
                      <Accordion.Toggle as={Button} variant="link" eventKey = {"pendingBookedLessons"}>
                        <div className="d-flex w-100 pt-3 justify-content-between no-gutters" id = {"pendingBookedLessons"}>
                            <h3>Pending Booked Lesson</h3>
                        </div>
                      </Accordion.Toggle>
                    </Card.Header>
                    <Accordion.Collapse eventKey={"pendingBookedLessons"}>
                      <Card.Body>
                        <Accordion>
                          {this.props.coursesList.map((course) => //per ogni mio corso
                            (this.isCourseSelected(course.courseName)) &&
                            <Card>
                              <Card.Header>
                                <Accordion.Toggle as={Button} variant="link" eventKey = {"course-" + course.courseName}>
                                  <CourseHeader course = {course.courseName}/>
                                </Accordion.Toggle>
                              </Card.Header>
                              <Accordion.Collapse eventKey={"course-" + course.courseName}>
                                <Card.Body>
                                  <ListGroup as="ul" variant="flush">
                                    <ListHeader />
                                    {this.props.waitingBookings.map((lesson) => 
                                      (course.courseId === lesson.courseId) &&
                                      (this.isLessonInDateBoundaries(lesson.startingTime, lesson.endingTime)) &&
                                      <LessonListItem key = {lesson.scheduleId} lesson = {lesson}
                                        updateSelectionMessage = {this.updateSelectionMessage}
                                        updateLessonSelectedState = {this.updateLessonSelectedState}
                                        selectLessonFunction = {this.props.selectLessonFunction}
                                        updateModalMessage={this.updateModalMessage}/>
                                    )}
                                  </ListGroup>
                                </Card.Body>
                              </Accordion.Collapse>
                            </Card>
                          )}
                        </Accordion>
                      </Card.Body>
                    </Accordion.Collapse>
                  </Card>
                }
                
              </Accordion>                
            </div>
            }
            {context.user && this.props.lessonsList && this.state.bookingCompleted &&
              <Modal show={this.state.bookingCompleted} animation={false}>
                <Modal.Header>
                  <Modal.Title>Booking response</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                  <div className="row-md-6">
                    {this.state.bookingMessage}
                  </div>
                </Modal.Body>
                <Modal.Footer>
                  <Button variant="secondary" type="button" 
                    onClick={(event) => {
                      event.preventDefault();
                      this.setState({bookingCompleted: false, bookingMessage: undefined});
                    }}>Close</Button>
                </Modal.Footer>
              </Modal>
            }
            {context.user && !this.props.lessonsList &&
              <NoItemsImage/>
            }
            {!context.user && <Redirect to="/login"/>}
          </>
        )}
      </AuthContext.Consumer>
    )
  }
}

function ListHeader() {
  return(
    <ListGroup.Item id = {"lessonList-header"}>
        <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
          <div className="col-sm-3">
              <h4>Starting Time</h4>
          </div>
          <div className="col-sm-3">
              <h4>Ending Time</h4>
          </div>
          <div className="col-sm-3">
              <h4># of Booked Seats</h4>
          </div>
          <div className="col-sm-3">
          </div>
        </div>
    </ListGroup.Item>
  );
}

function CourseHeader(props) {
  return(
    <div className="d-flex w-100 pt-3 justify-content-between no-gutters" id = {"course-" + props.course}>
        <h4>{props.course}</h4>
    </div>
  );
}
function NoItemsImage(props){
    return(
        <div className="col-sm-12 pt-3">
            <img width="800" height="600" className="img-button" src='./images/no_data_image.png' alt=""/>
        </div>
    );
}

export default lessonsList;
