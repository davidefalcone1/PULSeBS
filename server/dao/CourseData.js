//usata per la retrieve della lista dei corsi
class CourseData{
    constructor(courseId, courseName, teacherId) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
    }
}

module.exports = CourseData;