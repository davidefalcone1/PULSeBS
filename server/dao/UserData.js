class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

}

module.exports = UserData;