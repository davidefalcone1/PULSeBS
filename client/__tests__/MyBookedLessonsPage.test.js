import React from 'react';
import MyBookedLessonsPage from '../src/components/MyBookedLessonsPage';
import UserData from '../src/API/UserData';
import LessonData from '../src/API/LessonData';
import CourseData from '../src/API/CourseData';
import moment from 'moment';

Date.now = jest.fn(() => 1482363367071);//Avoid dynamic content on the page

describe('MyBookedLessonsPage', ()=>{
    let component;
    beforeAll(()=>{
        component = mount(
            <MyBookedLessonsPage courses={[new CourseData('XY123', 'course1', '654321', 2, 3, 5, 6, 7, 3, 5, 7), new CourseData('XY122', 'course2', '654321', 2, 3, 5, 6, 7, 3, 5, 7)]} lessonsList={[new LessonData(1, 'XY123', moment().add(1, 'days'), moment().add(1, 'days').add(1, 'hours'), 9, 90, false, true, 'A1', 100, 4, 33, 0)]} waitingBookings={[new LessonData(2, 'XY123', moment().add(2, 'days'), moment().add(2, 'days').add(1, 'hours'), 9, 90, false, true, 'A1', 100, 4, 33, 0)]} selectLessonFunction={()=>{}} updateMyBookedLessonsList={()=>{}}/>,
            {
                context: {user: new UserData(1, '12345', 'Davide Falcone', 'davidefalcone@polito.it', true)}
            }
        );
    });
    test('Renders correctly', ()=>{
        expect(component).toContainMatchingElement('#course-course1');
        expect(component).toContainMatchingElement('#course-course2');
    });
    test('filter', ()=>{
        const form = component.find('#lessonFilterForm').at(0);
        form.find('select').simulate('change', {target: {value: 'course1'}});
        form.find('select').simulate('change', {target: {value: 'Select course name'}});
    });
    test('accordion booked lessons', (done)=>{
        component.find('#bookedLessons').simulate('click');
        setTimeout(()=>{
            component.update();
            expect(component).toContainMatchingElement('.collapse.show');
            done();
        }, 500);
    });
    test('accordion pending lessons', (done)=>{
        component.find('#pendingBookedLessons').simulate('click');
        setTimeout(()=>{
            component.update();
            expect(component).toContainMatchingElement('.collapse.show');
            done();
        }, 500);
    });
    test('open calendar', ()=>{
        component.find('img').at(1).simulate('click');
        expect(component).not.toContainMatchingElement('.accordion');
    })
});