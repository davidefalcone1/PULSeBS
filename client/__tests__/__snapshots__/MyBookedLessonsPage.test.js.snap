// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyBookedLessonsPage Renders correctly 1`] = `
<div
  style={
    Object {
      "padding": "15px",
    }
  }
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h5>
        Filter your lessons!
      </h5>
    </div>
    <div
      className="text-right col"
      style={
        Object {
          "marginRight": "15px",
        }
      }
    >
      <img
        alt=""
        className="img-button"
        height="30"
        onClick={[Function]}
        src="./images/list_icon.png"
        width="30"
      />
      <span>
           
      </span>
      <img
        alt=""
        className="img-button"
        height="30"
        onClick={[Function]}
        src="./images/calendar_icon.png"
        width="30"
      />
    </div>
  </div>
  <form
    action=""
    className=""
    id="lessonFilterForm"
    method="POST"
    onSubmit={[Function]}
  >
    <div
      className="form-row"
    >
      <div
        className="form-group col"
      >
        <label
          className="control-label form-label"
        >
          Course Name
        </label>
        <select
          className="custom-select"
          name="courseName"
          onChange={[Function]}
          value="Select course name"
        >
          <option>
            Select course name
          </option>
          <option>
            Software engineering 2
          </option>
        </select>
      </div>
      <div
        className="form-group col"
      >
        <label
          className="control-label form-label"
        >
          Starting Day
        </label>
        <input
          className="form-control"
          min="2016-12-22"
          name="startingDay"
          onChange={[Function]}
          type="date"
          value=""
        />
      </div>
      <div
        className="form-group col"
      >
        <label
          className="control-label form-label"
        >
          Ending day
        </label>
        <input
          className="form-control"
          min="Invalid date"
          name="endingDay"
          onChange={[Function]}
          type="date"
          value=""
        />
      </div>
    </div>
    <div
      className="form-group"
    >
      <div>
        <button
          className="btn btn-secondary"
          onClick={[Function]}
          type="button"
        >
          Reset Fields
        </button>
      </div>
    </div>
  </form>
  <br />
  <div
    className="accordion"
  >
    <div
      className="card"
    >
      <div
        className="card-header"
      >
        <button
          className="btn btn-link"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <div
            className="d-flex w-100 pt-3 justify-content-between no-gutters"
            id="bookedLessons"
          >
            <h3>
              Booked Lessons
            </h3>
          </div>
        </button>
      </div>
      <div
        aria-expanded={null}
        className="collapse"
      >
        <div
          className="card-body"
        >
          <div
            className="accordion"
          >
            <div
              className="card"
            >
              <div
                className="card-header"
              >
                <button
                  className="btn btn-link"
                  disabled={false}
                  onClick={[Function]}
                  type="button"
                >
                  <div
                    className="d-flex w-100 pt-3 justify-content-between no-gutters"
                    id="course-Software engineering 2"
                  >
                    <h4>
                      Software engineering 2
                    </h4>
                  </div>
                </button>
              </div>
              <div
                aria-expanded={null}
                className="collapse"
              >
                <div
                  className="card-body"
                >
                  <ul
                    className="list-group list-group-flush"
                    onKeyDown={[Function]}
                  >
                    <div
                      className="list-group-item"
                      data-rb-event-key={null}
                      disabled={false}
                      id="lessonList-header"
                      onClick={[Function]}
                    >
                      <div
                        className="d-flex w-100 pt-3 justify-content-between no-gutters"
                      >
                        <div
                          className="col-sm-3"
                        >
                          <h4>
                            Starting Time
                          </h4>
                        </div>
                        <div
                          className="col-sm-3"
                        >
                          <h4>
                            Ending Time
                          </h4>
                        </div>
                        <div
                          className="col-sm-3"
                        >
                          <h4>
                            # of Booked Seats
                          </h4>
                        </div>
                        <div
                          className="col-sm-3"
                        />
                      </div>
                    </div>
                    <div
                      className="list-group-item"
                      data-rb-event-key={null}
                      disabled={false}
                      id="lesson-undefined"
                      onClick={[Function]}
                    >
                      <div
                        className="d-flex w-100 pt-3 justify-content-between no-gutters"
                      >
                        <div
                          className="col-sm-3"
                        >
                          <p
                            id="startDateOfLesson1"
                          >
                            Fri 23-12-2016 00:36
                          </p>
                        </div>
                        <div
                          className="col-sm-3"
                        >
                          <p
                            id="endDateOfLesson1"
                          >
                            Fri 23-12-2016 01:36
                          </p>
                        </div>
                        <div
                          className="col-sm-3"
                        >
                          <p
                            id="seatsStatusOfLesson1"
                          >
                            9
                             / 
                            90
                          </p>
                        </div>
                        <div
                          className="col-sm-3"
                        >
                          <button
                            className="btn btn-danger"
                            disabled={false}
                            id="deleteFieldOfLesson1"
                            onClick={[Function]}
                            type="button"
                          >
                            DELETE
                          </button>
                        </div>
                      </div>
                    </div>
                  </ul>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
